src/cargo/sources/git/utils.rs

```diff
@@ -694,12 +695,40 @@ pub fn with_fetch_options(
     let mut progress = Progress::new("Fetch", config);
     network::with_retry(config, || {
         with_authentication(url, git_config, |f| {
+            let mut last_recv = 0.0; // in Byte
+            let mut last_rate = 0.0; // in Byte/s
+            let mut last_update = Instant::now();
             let mut rcb = git2::RemoteCallbacks::new();
             rcb.credentials(f);
-
             rcb.transfer_progress(|stats| {
+                let indexed_deltas = stats.indexed_deltas();
+                let msg = if indexed_deltas > 0 {
+                    // Resolving deltas.
+                    format!(" ({}/{})", indexed_deltas, stats.total_deltas())
+                } else {
+                    // Receiving objects.
+                    let duration = last_update.elapsed();
+                    let (recv, rate) = if duration > Duration::from_secs(1) {
+                        let recv = stats.received_bytes() as f32;
+                        let rate = (recv - last_recv) / duration.as_secs_f32();
+                        last_recv = recv;
+                        last_rate = rate;
+                        last_update = Instant::now();
+                        (recv, rate)
+                    } else {
+                        (last_recv, last_rate)
+                    };
+                    fn format_bytes(bytes: f32) -> (&'static str, f32) {
+                        static UNITS: [&str; 5] = ["", "K", "M", "G", "T"];
+                        let i = (bytes.log2() / 10.0).min(4.0) as usize;
+                        (UNITS[i], bytes / 1024_f32.powi(i as i32))
+                    }
+                    let (rate_unit, rate) = format_bytes(rate);
+                    let (unit, recv) = format_bytes(recv);
+                    format!(" | {:.2}{}iB | {:.2}{}iB/s", recv, unit, rate, rate_unit)
```

Comment:
I tried this locally and I found this a little confusing where the first number here is the amount of data received but we unfortunately don't have an indicator of how much more data will be received or how much more data is remaining. I think if all we can show is the amount of total data received so far it's probably fine to drop this.

Having the transfer rate printed though is quite nice!

Reply:
`git` itself also only shows received bytes.

```
remote: Compressing objects: 100% (7/7), done.
Receiving objects:   0% (6896/1453394), 2.60 MiB | 271.00 KiB/s
```

But you're right it is not informative if there is no remaining bytes to show. I'll change it to look like following:

```
    Updating crates.io index
       Fetch [==>                      ]  14.50%, 258.45KiB/s
```

Does it look better?

Reply:
Yeah I think a percentage looks better because it gives you a sense of progress, where given a sense of progress and a rate of increase that should be pretty good for a terse-ish display!